interface/src/app/signature/enriched_signatures.js

```diff
@@ -0,0 +1,203 @@
+      }
+    ).$promise;
+
+    // Promise that gets the total number of genes:
+    var p2 = Gene.get(
+      {},
+      function success(response) {
+        self.geneNum = response.meta.total_count;
+      },
+      function error(err) {
+        var message = errGen('Failed to get total gene number: ', err);
+        $log.error(message);
+        self.statusMessage = message + '. Please try again later.';
+      }
+    ).$promise;
+
+    var participations;
+    // Promise that gets all participation records that are related to
+    // the genes in URL:
+    var p3 = Participation.get(
+      {'related_genes': $stateParams.genes, 'limit': 0},
+      function success(response) {
+        participations = response.objects;
+      },
+      function error(err) {
+        var message = errGen(
+          'Failed to get signature-gene participations: ', err);
+        $log.error(message);
+        self.statusMessage = message + '. Please try again later.';
+      }
+    ).$promise;
+
+    var pValueSigDigits = 3;
+    // groupedGenes is an object that behaves like a 2-D table of genes.
+    // groupedGenes[participationTypeID][signatureID] is an object whose
+    // keys are gene IDs and values are genes names of those genes that
+    // participate with signatureID and with participationTypeID.
+    var groupedGenes = Object.create(null);
+
+    // Helper function that calculates the enrichment for each signature
+    // that includes at least one gene in the url.
+    var calculateEnrichments = function(typeName, genesBySignatures) {
+      var m = genesInUrl.length;
+      var N = self.geneNum;
+      var pValueArray = [];
+      var matchedGenesBySignature = [];
+      var signatureIDs = Object.keys(genesBySignatures);
+      signatureIDs.forEach(function(sigID) {
+        var genes = genesBySignatures[sigID];
+        var n = Object.keys(genes).length, k = 0;
+        var matchedGenes = [], selectedGene;
+        var pValue;
+        for (var i = 0; i < genesInUrl.length; i++) {
+          selectedGene = genesInUrl[i];
+          if (genes[selectedGene]) {
+            k++;
+            matchedGenes.push(genes[selectedGene]);
+          }
+        }
+        matchedGenesBySignature.push(matchedGenes);
+        // Four parameters are needed in the formula:
+        //     hyperGeometricTest(k, m, n, N)
+        // k: the number of heavy genes related to sigID in genesInUrl;
+        // m: the number of genes in genesInUrl;
+        // n: the total number of heavy genes related to sigID;
+        // N: the total number of genes.
+        pValue = 1 - MathFuncts.hyperGeometricTest(k, m, n, N);
+        pValueArray.push(pValue);
+      });
+
+      var significantSignatures = [];
+      var correctedPValues = MathFuncts.multTest.fdr(pValueArray);
+      correctedPValues.forEach(function(element, index) {
+        var correctedPValue = element.toPrecision(pValueSigDigits);
+        if (correctedPValue < self.pValueCutoff) {
```

Comment:
This test for the `pValueCutoff` should happen *before* the `.toPrecision()` is done. We should use all of the precision we have to make the cutoff determination, while the `.toPrecision(pValuesSigDigits)` can be thought of as "formatting" the output.

If we don't do so in the right order, rounding could bump some values above the cutoff inappropriately.

Reply:
Good point. I will do that.